### PR TITLE
Add HNClient TPR message request

### PR DESCRIPTION
**Pull request by TiagoGraf89**

### CODE REVIEW

**Comment by sslaws** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=medicationservice_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=hnclientservice_pr_109

**Comment by sslaws** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=medicationservice_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=hnclientservice_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=hnclientservice_pr_109

**Comment by sslaws** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=Medicationservice_pr_109

File: Apps/HNClient/src/Models/HNMessage.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// limitations under the License.
+//-------------------------------------------------------------------------
+namespace HealthGateway.HNClient.Models
+{
+    /// <summary>
+    /// A simple model object representing a response and error state from HNSecure.
+    /// </summary>
+    public class HNMessage
+    {
+        /// <summary>
```
[the remark this answers is not in the document]
Comment (TiagoGraf89) [reply Apps/HNClient/src/Models/HNMessage.cs]: We actually use the empty constructor to return errors

File: Apps/Medication/src/Controllers/MedicationController.cs [hunk 2]
```diff
@@ -56,7 +56,11 @@ public MedicationController(IMedicationService svc)
         [Route("{hdid}")]
         public async Task<List<MedicationStatement>> GetMedicationStatements(string hdid)
         {
-            return await this.service.GetMedicationStatementsAsync(hdid).ConfigureAwait(true);
+            // Uses hardcoded phn until we have the token setup.
+            // string phn = this.GetPatientPHN(hdid);
+            string phn = "0009735353315";
```
[the remark this answers is not in the document]
Comment (TiagoGraf89) [reply Apps/Medication/src/Controllers/MedicationController.cs]: To be implemented once we have a token in our med service

File: Apps/Medication/src/Controllers/MedicationController.cs [hunk 3]
```diff
@@ -56,7 +56,11 @@ public MedicationController(IMedicationService svc)
         [Route("{hdid}")]
         public async Task<List<MedicationStatement>> GetMedicationStatements(string hdid)
```
[the remark this answers is not in the document]
Comment (TiagoGraf89) [reply Apps/Medication/src/Controllers/MedicationController.cs]: no JWT yet! but we should

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=hnclientservice_pr_109

**Comment by sslaws** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=Medicationservice_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=hnclientservice_pr_109

**Comment by sslaws** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=Medicationservice_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=hnclientservice_pr_109

**Comment by sslaws** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=Medicationservice_pr_109

File: Apps/Medication/src/Parsers/TRPParser.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    using System.Collections.Generic;
+    using System.Globalization;
+    using HealthGateway.MedicationService.Models;
+    using HL7.Dotnetcore;
+    using Microsoft.Extensions.Configuration;
+
+    /// <summary>
+    /// Parser of TRP (Patient Profile) messages.
+    /// </summary>
+    public static class TrpParser
```
Comment (FuriousLlama) [inline Apps/Medication/src/Parsers/TRPParser.cs]: I dont believe this should be static. Instead the parser could be created on the fly as a worker and get dispose or we could set it as a dependency injection singleton.

File: Apps/Medication/src/Parsers/TRPParser.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        private const string TRACE = "101010";
+
+        /// <summary>
+        /// Creates a TRP request message to HNClient.
+        /// </summary>
+        /// <param name="phn">The patient phn.</param>
+        /// <param name="userId">The requester user id.</param>
+        /// <param name="ipAddress">The requester ip address phn.</param>
```
Comment (FuriousLlama) [inline Apps/Medication/src/Parsers/TRPParser.cs]: Is the ip address going to be the Webclient's one?

Comment (TiagoGraf89) [reply Apps/Medication/src/Parsers/TRPParser.cs]: it should be the client requesting the data (the ip from the end user)

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=hnclientservice_pr_109

File: Apps/Medication/src/Parsers/TRPParser.cs [hunk 6]
```diff
@@ -107,7 +105,7 @@ public static HNMessage CreateRequestMessage(string phn, string userId, string i
             zcc.AddNewField(string.Empty); // Relationship
             zcc.AddNewField(string.Empty); // Patient First Name
             zcc.AddNewField(string.Empty); // Patient Last Name
-            zcc.AddNewField(phn); // Provincial Health Care ID
+            zcc.AddNewField(id); // Provincial Health Care ID
```
Comment (FuriousLlama) [inline Apps/Medication/src/Parsers/TRPParser.cs]: IMO provincialHealthcareId is better than just ID as a parameter name

File: Apps/Medication/src/Parsers/IHnParser.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+namespace HealthGateway.MedicationService.Parsers
+{
+    using System.Collections.Generic;
+    using HealthGateway.MedicationService.Models;
+
+    /// <summary>
+    /// The HNClient parser interface.
+    /// </summary>
+    /// <typeparam name="T">The model type.</typeparam>
+    public interface IHnParser<T>
```
Comment (FuriousLlama) [inline Apps/Medication/src/Parsers/IHnParser.cs]: IHnParser -> IHNParser or IHNMessageParser

**Comment by sslaws** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=Medicationservice_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=commonlibrary_pr_109

**Comment by healthgatewaybc** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=hnclientservice_pr_109

**Comment by sslaws** [comment]
SonarQube Report: https://sonarqube-gateway.pathfinder.gov.bc.ca/dashboard?id=Medicationservice_pr_109

**Pull request merged**